compiler/AST/iterator.cpp

```diff
@@ -765,21 +937,45 @@ static void
 buildIncr(IteratorInfo* ii, BlockStmt* singleLoop) {
   BlockStmt* incrBody = new BlockStmt();
 
-  if (singleLoop != NULL && singleLoop->isCForLoop() == true) {
-    CForLoop*  cforLoop  = toCForLoop(singleLoop);
-    BlockStmt* incrBlock = NULL;
-
-    Symbol*    ic        = ii->advance->getFormal(1);
-    SymbolMap  map;
+  // See Notes #1 and #2.
+  SymbolMap map;
+  Symbol* advance_this = ii->advance->getFormal(1);
+  Symbol* my_this = ii->incr->getFormal(1);
+  INT_ASSERT(my_this == ii->incr->_this);
+  map.put(advance_this, my_this);
 
-    // In copied expressions, replace _ic with incr->_this .
-    map.put(ic, ii->incr->_this);
+  if (singleLoop != NULL) {
+    if (singleLoop->isCForLoop() == true) {
+      CForLoop*  cforLoop  = toCForLoop(singleLoop);
+      BlockStmt* incrBlock = NULL;
 
-    incrBlock = cforLoop->incrBlockGet();
+      incrBlock = cforLoop->incrBlockGet();
 
-    for_alist(expr, incrBlock->body) {
-      incrBody->insertAtTail(expr->copy(&map));
+      for_alist(expr, incrBlock->body) {
+        incrBody->insertAtTail(expr->copy(&map));
+      }
     }
+    else if (singleLoop->isForLoop())
```

Comment:
The ForLoop being replaced would not give rise to an incr clause, but the underlying iterator might.  I didn't want to distinguish how the underlying iterator is rendered, so I just avoided the question by fully implementing the interface.  If someone can prove that the init and incr wrappers can be empty, then the corresponding if (singleLoop->isForLoop()) clauses can be removed from the corresponding buildInit() and buildIncr() routines.
I think I have a counterexample, though I also am reaching the point where much additional effort on this single failing test case seems unwarranted.